### PR TITLE
Adding state as dropdown in Address Questions

### LINKED ISSUE #1485: Make Address question type's State input a dropdown
The Address question type currently uses a free-text-entry for the State field. The UX would be better (easier to use, harder to get wrong) if we instead used a dropdown.

image

In the dropdown, offer all 50 US states, plus DC, PR, American Samoa, Guam, Northern Mariana Islands, and the Virgin Islands (56 in total).

We may want to style each dropdown item as `$Abbreviation - $State`, such as `WA - Washington` or `CA - California`. This would make the list easier to skim.

### PR DESCRIPTION
**Pull request by swatkat1**
### Description

Converting the state field in the address question as a dropdown with the 50 states, DC and 8 territories.
Also fixing browser tests that are failing due to change from PDT to PST in screenshot comparison where timestamps are used.

## Release notes:

Adding state as dropdown in Address Questions. Two things to note here are- The name of the state will appear as "WA" or "WY" and the values it will have in the submitted form should be just WA.

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

#### User visible changes

- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [x] Manually tested at 200% size
- [ ] Manually evaluated tab order

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/application.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Issue(s) this completes

Fixes #1485 

Screenshot:
image

### CODE REVIEW

**Comment by shanemc-goog** [comment]
Please fix the timezone browser test issue in another PR. PRs should be focused as much as possible and it clouds this one, as well masks the fix that it necessary for that issue.

**Review by shanemc-goog — CHANGES_REQUESTED** [review]
Looks good:

1. Please evaluate and check off "Manually evaluated tab order" for a11y

2. Please merge in #3769 and update the image diffs

File: browser-test/src/support/applicant_questions.ts [hunk 1]
```diff
@@ -36,17 +38,18 @@ export class ApplicantQuestions {
     await this.page.waitForSelector('.cf-address-street-1 input')
     await this.page.waitForSelector('.cf-address-street-2 input')
     await this.page.waitForSelector('.cf-address-city input')
-    await this.page.waitForSelector('.cf-address-state input')
```
Comment (gwendolyngoetz) [inline browser-test/src/support/applicant_questions.ts:39]: Can you see if using `.cf-address-state select` will allow you to keep this line?

File: browser-test/src/support/applicant_questions.ts [hunk 2]
```diff
@@ -36,17 +38,18 @@ export class ApplicantQuestions {
...
     // Check values are equal to expected
     await this.validateInputValue(street, '.cf-address-street-1 input')
     await this.validateInputValue(line2, '.cf-address-street-2 input')
     await this.validateInputValue(city, '.cf-address-city input')
-    await this.validateInputValue(state, '.cf-address-state input')
+    await this.validateDropdownValue(state, '.cf-address-state')
     await this.validateInputValue(zip, '.cf-address-zip input')
   }
-
+  async validateDropdownValue(value: string, dropdownId: string) {
```
Comment (gwendolyngoetz) [inline browser-test/src/support/applicant_questions.ts:50]: Nit: Can you move this down to be with the other methods that validate input so they stay together?

File: server/app/views/questiontypes/AddressQuestionRenderer.java [hunk 3]
```diff
@@ -69,16 +71,23 @@ protected DivTag renderInputTags(
...
+    SelectWithLabel stateField =
+        (SelectWithLabel)
+            new SelectWithLabel()
+                .setFieldName(addressQuestion.getStatePath().toString())
+                .setValue(addressQuestion.getStateValue().orElse(""))
+                .setLabelText(messages.at(MessageKey.ADDRESS_LABEL_STATE.getKeyName()))
+                .setOptionGroups(
+                    ImmutableList.of(
+                        SelectWithLabel.OptionGroup.builder()
+                            .setLabel("Pick your State")
```
Comment (gwendolyngoetz) [inline server/app/views/questiontypes/AddressQuestionRenderer.java:83]: Does this string need to be localized?

File: browser-test/src/support/applicant_questions.ts (hunk 1 above)
Comment (swatkat1) [reply browser-test/src/support/applicant_questions.ts:39]: Tried. It didn't work.

File: browser-test/src/support/applicant_questions.ts (hunk 2 above)
Comment (swatkat1) [reply browser-test/src/support/applicant_questions.ts:50]: sure

File: server/app/views/questiontypes/AddressQuestionRenderer.java (hunk 3 above)
Comment (swatkat1) [reply server/app/views/questiontypes/AddressQuestionRenderer.java:83]: Good catch! Will do.

**Comment by gwendolyngoetz** [comment]
@swatkat1  I was looking at the screenshot and noticed the height of the state dropdown needs some adjustment to match the other controls.

**Comment by swatkat1** [comment]
> Looks good:
> 
>     1. Please evaluate and check off "Manually evaluated tab order" for a11y
> 
>     2. Please merge in #3769 and update the image diffs

@shanemc-goog  - Should i revert the images files?

File: server/app/views/questiontypes/AddressQuestionRenderer.java [hunk 4]
```diff
@@ -107,10 +117,76 @@ protected DivTag renderInputTags(
                 addressOptionalField.getInputTag(),
                 /** Third line of address entry: City, State, Zip */
                 div()
-                    .withClasses("grid", "grid-cols-3", "gap-2")
+                    .withClasses("grid", "grid-cols-3", "gap-3")
                     .with(
-                        cityField.getInputTag(), stateField.getInputTag(), zipField.getInputTag()));
+                        cityField.getInputTag(),
+                        stateField.getSelectTag(),
```
Comment (shanemc-goog) [inline server/app/views/questiontypes/AddressQuestionRenderer.java:123]: Add a TODO here with the issue you made to fix the styling

**Comment by shanemc-goog** [comment]
Suggested fixes for the description

"as a dropdown with 56 states and their state codes." -> "as a dropdown with the 50 states and 9 territories.

The release notes still mentions the state name

File: server/app/views/questiontypes/AddressQuestionRenderer.java (hunk 4 above)
Comment (swatkat1) [reply server/app/views/questiontypes/AddressQuestionRenderer.java:125]: Done!

**Review by shanemc-goog — APPROVED** [review]

**Pull request merged**